feat(discover): Introduce a way to build snql queries

wmak opened the pull request:
- The plan here is to replace the current calls to `get_filter` and `resolve_field_list` with this new `QueryBuilder` class eventually and then use it for `discover.query`
- Currently most fields/filters/functions aren't implemented yet, but they should mostly be pretty easy to convert from the existing code
  - that is, this is a base so that we can start converting the beasts that are `search/event/fields.py` and `search/event/filter.py` over incrementally
  - In the meantime I **intentionally** have **not** added a way to run the query in discover yet, this is cause I don't want to accidentally have a user get an error like `not implemented in snql filter parsing yet`
- ~I'm also concerned that this builder.py file will grow excessively like api/event_search.py did before it was split up, I'm open to suggestions to try to keep it cleaner~ I've split builder into QueryFilter + QueryFields

wmak committed:
feat(discover): Introduce a way to build snql queries

- The idea here is to replace the current calls to `get_filter` and
  `resolve_field_list` with this new `QueryBuilder` class eventually and
  then use it for discover.query
- Currently most fields/filters/functions aren't implemented yet, but
  they should mostly be pretty easy to convert from the existing code
  - In the meantime i've *intentionally* not added a way to run the
    query inf discover yet, this is cause I don't want to accidentally
    have a user get an error like "this isn't supported by snql"
- I'm also concerned that this builder.py file will grow excessively
  like api/event_search.py did before it was split up, open to
  suggestions to try to keep it cleaner

evanh commented on src/sentry/search/events/builder.py:
```diff
@@ -0,0 +1,272 @@
+
+# TODO: this should be a TypedDict instead
+ParamsType = Mapping[str, Union[List[int], int, str, datetime]]
+AggregateType = Union[Function, CurriedFunction]
```
Function is a subclass of CurriedFunction, FYI.

evanh commented on src/sentry/search/events/builder.py:
```diff
@@ -0,0 +1,272 @@
+
+        # We want to use group_id elsewhere so shouldn't be removed from the dataset
+        # but if a user has a tag with the same name we want to make sure that works
+        if name in {"group_id"}:
```
```suggestion
        if name == "group_id":
```

evanh commented on src/sentry/search/events/builder.py:
```diff
@@ -0,0 +1,272 @@
+WhereType = Union[Condition, BooleanCondition]
+SelectType = Union[Column, CurriedFunction, Function]
+
+OPERATOR_MAP: Mapping[str, Op] = {
```
You could cast the operators instead of maintaining this map. `Op("=") == Op.EQ`

wmak replied:
_whoaa_ nice 👏

evanh commented on src/sentry/search/events/builder.py:
```diff
@@ -0,0 +1,272 @@
+        if len(validated) == len(self.orderby_columns):
+            return validated
+
+        raise InvalidSearchQuery("Cannot order by a field that is not selected.")
```
You can actually. I made this error when I was writing the SnQL parser.

wmak replied:
Gonna leave this as-is for now to keep functionality consistent, but I'll add a comment explaining that this error is no longer valid.

evanh commented on src/sentry/search/events/builder.py:
```diff
@@ -0,0 +1,272 @@
+        self.where: List[WhereType] = []
+
+        self.params = params
+        self.dataset = dataset
```
I wonder if it would be cleaner to have a `self.query: Query` that you call the `set_*` functions on instead of having `self.where`, `self.limit` etc. I'm not saying you need to change this, but something to consider.

wmak replied:
👍 Yeah I had considered that too, but since Query is immutable so I figured this way its easier to edit the clauses at will and create one Query instance at the very end.

evanh commented on tests/sentry/search/events/test_builder.py:
```diff
@@ -0,0 +1,124 @@
+            Condition(Column("email"), Op.EQ, "foo@example.com"),
+            Condition(Column("release"), Op.EQ, "1.2.1"),
+            *self.default_conditions,
+        ]
```
I would probably do `query.get_snql_query()` to make sure what you're building is actually valid.

wmak replied:
:+1: that makes sense, should I do `query.get_snql_query().validate()` too? Or is instantiating the `Query` sufficient?

wmak replied:
nvm answered my own q, gotta call the validate too, the super simple tests fail cause of 
```
E   snuba_sdk.query_visitors.InvalidQuery: query must have at least one expression in select
```

evanh reviewed (commented):
I like where this is going, but I have some questions.

Have you thought about replacing the `SearchVisitor` class when parsing the conditions? For example, parse a query directly into a `Condition` instead of the intermediate `SearchFilter`. You wouldn't need to rewrite the grammar at all, instead return different data structures from the `visit_*` functions in the parsing class. That would simplify the logic of this class quite a bit.

wmak commented:
> Have you thought about replacing the SearchVisitor class when parsing the conditions? For example, parse a query directly into a Condition instead of the intermediate SearchFilter.

Oh this is a great idea, but I think this might be better as a follow up once we can get rid of the existing parsing for the json syntax so that we can keep running these two things in parallel.

wmak committed:
ref: Addressing PR comments

- Simplifying type for AggregateType
- Cast operators instead of maintaining a map

wmak committed:
ref: Validating built queries too

- Selected columns is not a required variable, since I can forsee us
  using the QueryBuilder to parse a query and then setting the
  aggregate/columns by hand for more custom behaviour

evanh commented:
> I'm also concerned that this builder.py file will grow excessively like api/event_search.py did before it was split up, I'm open to suggestions to try to keep it cleaner

Part of this problem was that there was a lot of fighting the JSON syntax, where a lot of code was complicated just because the syntax was complicated. I am _trying_ to fix that with the SDK, but if you ever feel like "man I wish the SDK did X, it would make this code a lot cleaner" I want to hear about it. I don't want to switch one pain point for another.

wmak committed:
ref: Splitting QueryBuilder into two classes

- This is so that QueryBuilder doesn't grow ridiculously large

wmak committed:
ref: move allowlist to constants, fix: missplaced comment

wmak committed:
Merge branch 'master' into feat/wmak/snql-query-builder

wmak commented on src/sentry/search/events/fields.py:
```diff
@@ -1604,3 +1611,59 @@ def is_accessible(self, acl=None):
+        if len(validated) == len(self.orderby_columns):
+            return validated
+
+        # TODO: This is no longer true, can order by fields that aren't selected, keeping
+        # for now so we're consistent with the existing functionality
```
:+1: Yep https://github.com/getsentry/sentry/pull/26286#discussion_r641736675

evanh replied:
It was always valid from a Clickhouse perspective, and I think possible with the legacy snuba query as well.

Zylphrex replied:
I see, let's leave it for consistency for now then.

wmak commented on src/sentry/search/events/types.py:
```diff
@@ -0,0 +1,13 @@
+WhereType = Union[Condition, BooleanCondition]
+# TODO: this should be a TypedDict instead
+ParamsType = Mapping[str, Union[List[int], int, str, datetime]]
+# Function is a subclass of CurriedFunction
+AggregateType = Union[CurriedFunction]
+SelectType = Union[Column, Function, CurriedFunction]
```
Sigh unfortunately not:
```python
[E] Argument of type "List[SelectType] | None" cannot be assigned to parameter "select" of type "List[Column | CurriedFunction | Function] | None"
```

but AggregateType is probably redundant since it should just be `CurriedFunction` instead of `Union[CurriedFunction]` gonna continue using the alias for readability though with `SelectType`

wmak commented on tests/sentry/search/events/test_builder.py:
```diff
@@ -0,0 +1,121 @@
+            Column("email"),
+            Column("release"),
+        ]
+        query.get_snql_query().validate()
```
I prefer asserting that the conditions we constructed are valid first before validating the snql query. This is cause I think it's clearer what to fix if you see the select changed (in this example), rather than having to figure out why your query is invalid

wmak committed:
ref: Adressing PR comments

- have format_search_filter return a condition rather than change where
  for you
- invalid syntax for the group_id condition
- removing redundant AggregateType

wmak commented:
> Part of this problem was that there was a lot of fighting the JSON syntax, where a lot of code was complicated just because the syntax was complicated. I am trying to fix that with the SDK, but if you ever feel like "man I wish the SDK did X, it would make this code a lot cleaner" I want to hear about it. I don't want to switch one pain point for another.

100% and so far SnQL looks to be a really nice upgrade. To be clear though the original comment here is aimed at how `api/event_search.py` became the dumping ground for all our query logic so it was getting hard to maintain, and I wanted to avoid that happening to `search/events/builder.py` here.

markstory commented on src/sentry/search/events/filter.py:
```diff
@@ -764,3 +768,130 @@ def format_search_filter(term, params):
+
+    def resolve_where(self, query: str) -> None:
+        try:
+            parsed_terms = parse_search_query(query, allow_boolean=True, params=self.params)
```
If this did translate directly to SNQL objects you might have a tricky situation with Issue search which use this function as well.

evanh replied:
I was thinking of having two subclasses of NodeVisitor. Then different parts of Sentry could choose whether they want the grammar to give them SnQL objects or legacy objects based on the class they use.

evanh commented on src/sentry/search/events/fields.py:
```diff
@@ -1604,3 +1611,59 @@ def is_accessible(self, acl=None):
+            if isinstance(field, SnqlColumn) and field not in self.columns:
+                self.columns.append(field)
+
+    def resolve_field(self, field) -> Union[SnqlColumn, SnqlFunction]:
```
```suggestion
    def resolve_field(self, field: str) -> Union[SnqlColumn, SnqlFunction]:
```

evanh commented on src/sentry/search/events/fields.py:
```diff
@@ -1604,3 +1611,59 @@ def is_accessible(self, acl=None):
+        if not isinstance(field, str):
+            raise InvalidSearchQuery("Field names must be strings")
+
+        match = is_function(field)
```
A lot of this parsing code (functions, orderby etc.) is pretty close to what the SDK does in the `legacy` package. Maybe you could reuse some of those? I'm hoping ultimately to deprecate that entire package, but if leveraging it helps with the transition to SnQL then I think it would be fine.

wmak replied:
Hmmm would it be better to base our implementation on the `legacy` package if needed instead of using it here, that way we don't need to release/update the sdk if there's a need for us to change anything? (and so you can still deprecate it as needed)

evanh replied:
I think if the functionality in `legacy` is going to be useful generally, then I just won't deprecate it.

This opens up a larger question actually: When I wrote the SDK I had thought that a class like `QueryBuilder` wouldn't be necessary. That the SDK Query class would be all that was needed. It seems like that assumption wasn't really true though, since there is logic needed in an intermediary class like `QueryBuilder`. My question is: do you think that you will ever _not_ need the `QueryBuilder` class? Does the Query data structure from the SDK actually fit the needs you have?

fpacifici replied:
If we keep the `legacy` package around could we move it out of the sdk? 
Reasons:
- that package contains a number of hacks meant for the rollout. there is no clean way to fully transform a legacy query into a SnQL query mainly because the original legacy grammar is inherently ambiguous. 
- it contains features that are not complete and only meant to work in the specific cases we encountered during the rollout. That is not what you would like in a generic api. If that solves a specific product need, then it should be in the product.
- Portability: if we wrote a nodeJS or rust service that had to query snuba we would have to port the sdk. Then we would have to either port the legacy package as well, thus carrying over the technical debt it contain on other platforms or make the port partial, which will be a pain to use as the api provided would not be the same the python sdk provides.

fpacifici replied:
But also, to echo Evan, What's the gap between the sdk Query class and the QueryBuilder ? Is it just a matter of the syntax used by the user to provide the query? Then a translation layer somewhere is inevitable and it would be product specific (other areas would not need it). In such case is there a reason to have an intermediate representation between the parser and SnQL?

wmak replied:
So the QueryBuilder class exists more as a convenient way to maintain parameters/state while parsing functions/fields. The main motivation for it is the current `params` function parameter on the majority of the functions in `search/events/fields.py` and `search/events/functions.py`.

The class doesn't necessarily have to continue existing and could be replaced with a series of functions like we do today.

evanh commented on src/sentry/search/events/filter.py:
```diff
@@ -764,3 +768,130 @@ def format_search_filter(term, params):
+        if "start" not in self.params or "end" not in self.params:
+            raise InvalidSearchQuery("Cannot query without a valid date range")
+
+        self.where.append(Condition(self.column("timestamp"), Op.GTE, self.params["start"]))
```
These params should all be validated. `start` and `end` should be datetimes, `project_id` should be ints etc.

evanh commented on src/sentry/search/events/types.py:
```diff
@@ -0,0 +1,11 @@
+from snuba_sdk.function import CurriedFunction, Function
+
+WhereType = Union[Condition, BooleanCondition]
+# TODO: this should be a TypedDict instead
```
```suggestion
# TODO: this should be a dataclass instead
```
(nit)

wmak committed:
ref: Addressing PR comments

- change todos on params to mention dataclass
- validate params on query construction

wmak committed:
Merge branch 'master' into feat/wmak/snql-query-builder

evanh reviewed (approved):
Excited to see this moving forwards! Well done!

Zylphrex reviewed (approved):
Can't wait for snql

wmak committed:
Merge branch 'master' into feat/wmak/snql-query-builder
